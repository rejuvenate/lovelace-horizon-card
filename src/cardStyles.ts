import { css } from 'lit'

export default css`
  .horizon-card {
    --horizon-card-primary: var(--primary-text-color, #000000);
    --horizon-card-secondary: var(--secondary-text-color, #828282);
    --horizon-card-accent: #d7d7d7;

    --horizon-card-lines: var(--horizon-card-accent);
    --horizon-card-field-name-color: var(--horizon-card-secondary);
    --horizon-card-field-value-color: var(--horizon-card-primary);

    --horizon-card-stop-invisible: rgb(0,0,0,0);
    --horizon-card-stop-sun-color: #f9d05e;
    --horizon-card-stop-dawn-color: #393b78;
    --horizon-card-stop-day-color: #8ebeeb;
    --horizon-card-stop-dusk-color: #393b78;

    padding: 0.5rem;
    font-size: 1.3rem;
    font-family: var(--primary-font-family);
  }

  .horizon-card.horizon-card-dark {
    --horizon-card-primary: #ffffff;
    --horizon-card-secondary: #828282;
    --horizon-card-accent: #464646;
  }

  .horizon-card-field-row {
    display: flex;
    justify-content: space-around;
    margin: 1rem 1.5rem 0 1.5rem;
  }

  .horizon-card-text-container {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .horizon-card-field-name {
    color: var(--horizon-card-field-name-color);
  }

  .horizon-card-field-value {
    color: var(--horizon-card-field-value-color);
  }

  .horizon-card-header {
    display: flex;
    justify-content: space-between;
    margin: 0 4rem 0 4rem;
  }

  .horizon-card-header .horizon-card-field-value {
    font-size: 1.85rem;
  }

  .horizon-card-title {
    margin: 0.5rem 0.5rem 3rem 3rem;
    font-size: 2rem;
    font-weight: 500;
    color: var(--horizon-card-primary);
  }

  .horizon-card-graph {
    shape-rendering="geometricPrecision";
    margin: 2rem 0 2rem 0;
  }

  .horizon-card-graph .sunInitialStop {
    stop-color: var(--horizon-card-stop-sun-color);
  }

  .horizon-card-graph .sunMiddleStop {
    stop-color: var(--horizon-card-stop-sun-color);
  }

  .horizon-card-graph .sunEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .dawnInitialStop {
    stop-color: var(--horizon-card-stop-dawn-color);
  }

  .horizon-card-graph .dawnMiddleStop {
    stop-color: var(--horizon-card-stop-dawn-color);
  }

  .horizon-card-graph .dawnEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .dayInitialStop {
    stop-color: var(--horizon-card-stop-day-color);
  }

  .horizon-card-graph .dayMiddleStop {
    stop-color: var(--horizon-card-stop-day-color);
  }

  .horizon-card-graph .dayEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .duskInitialStop {
    stop-color: var(--horizon-card-stop-dusk-color);
  }

  .horizon-card-graph .duskMiddleStop {
    stop-color: var(--horizon-card-stop-dusk-color);
  }

  .horizon-card-graph .duskEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .card-config ul {
    list-style: none;
    padding: 0 0 0 1.5rem;
  }

  .card-config li {
    padding: 0.5rem 0;
  }
`
