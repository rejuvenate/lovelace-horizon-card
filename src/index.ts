import './components/horizonCard'
import './components/horizonCardEditor'

