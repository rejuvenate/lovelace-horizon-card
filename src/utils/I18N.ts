import { formatNumber, FrontendLocaleData, LocalizeFunc, NumberFormat, TimeFormat } from 'custom-card-helpers'

import { Constants } from '../constants'
import { THorizonCardI18NKeys } from '../types'

export class I18N {
  private readonly localization: THorizonCardI18NKeys
  private readonly dateFormatter: Intl.DateTimeFormat
  private readonly locale: FrontendLocaleData
  private readonly localizeFunc: LocalizeFunc

  constructor (language: string, timeZone: string, timeFormat: TimeFormat, numberFormat: NumberFormat,
               localizeFunc: LocalizeFunc) {
    this.localization = I18N.matchLanguageToLocalization(language)

    this.dateFormatter = I18N.createDateFormatter(language, timeZone, timeFormat)

    this.locale = {
      language,
      time_format: timeFormat,
      number_format: numberFormat
    }

    this.localizeFunc = localizeFunc
  }

  public formatDateAsTime (date: Date): string {
    let time = this.dateFormatter.format(date)
    if (this.locale.language === 'bg') {
      // Strips " ч." from times in Bulgarian - some major browsers insist on putting it there:
      // https://unicode-org.atlassian.net/browse/CLDR-11545
      // https://unicode-org.atlassian.net/browse/CLDR-15802
      time = time.replace(' ч.', '')
    }
    return time
  }

  public formatDecimal (decimal: number): string {
    return formatNumber(decimal, this.locale)
  }

  /**
   * TR -> TRanslation
   * @param translationKey The key to lookup a translation for
   * @returns The string specified in the translation files
   */
  public tr (translationKey: string): string {
    // if the translation isn't completed in the target language, fall back to english
    // give ugly string for developers who misstype
    return this.localization[translationKey] ?? Constants.FALLBACK_LOCALIZATION[translationKey]
      ?? `Translation key '${translationKey}' doesn't have a valid translation`
  }

  public localize (key: string): string {
    return this.localizeFunc(key)
  }

  private static matchLanguageToLocalization (language: string) {
    let data = Constants.LOCALIZATION_LANGUAGES[language]
    if (data === undefined) {
      // Matches things like en-GB to en, es-419 to es, etc.
      data = Constants.LOCALIZATION_LANGUAGES[language.split('-', 2)[0]]
    }
    if (data === undefined) {
      data = Constants.FALLBACK_LOCALIZATION
    }

    return data
  }

  private static createDateFormatter (language: string, timeZone: string, timeFormat: TimeFormat): Intl.DateTimeFormat {
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat
    const dateTimeFormatOptions: Intl.DateTimeFormatOptions = {
      hour: 'numeric',
      minute: '2-digit',
      timeZone: timeZone
    }

    // mimics home assistant's logic
    if (timeFormat === 'language' || timeFormat === 'system') {
      const testLanguage = timeFormat === 'language' ? language : undefined
      const test = new Date().toLocaleString(testLanguage)
      dateTimeFormatOptions.hour12 = test.includes('AM') || test.includes('PM')
    } else {
      // Casting to string allows both "time_format: 12" and "time_format: '12'" in YAML
      dateTimeFormatOptions.hour12 = String(timeFormat) === '12'
    }

    let timeLocale = language
    if (!dateTimeFormatOptions.hour12) {
      // Prevents times like 24:00, 24:15, etc. with the 24h clock in some locales.
      // Home Assistant does this only for 'en' but zh-Hant for example suffers from the same problem.
      timeLocale += '-u-hc-h23'
    }
    return new Intl.DateTimeFormat(timeLocale, dateTimeFormatOptions)
  }
}
