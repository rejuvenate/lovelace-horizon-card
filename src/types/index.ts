import { LovelaceCardConfig } from 'custom-card-helpers'

import { I18N } from '../utils/I18N'

export type THorizonCardFields = {
  sunrise?: boolean
  sunset?: boolean

  dawn?: boolean
  noon?: boolean
  dusk?: boolean

  azimuth?: boolean
  elevation?: boolean
}

export interface IHorizonCardConfig extends LovelaceCardConfig {
  i18n?: I18N
  darkMode?: boolean
  language?: string

  title?: string,
  component?: string
  use12hourClock?: boolean

  fields?: THorizonCardFields
}

export interface IConfigChangedEvent <T> extends CustomEvent<T> {
  target: CustomEvent['target'] & {
    configValue: string
    selected?: string
    checked?: boolean
  }
}

export type TSunInfo = {
  sunrise: number,
  sunset: number

  dawnProgressPercent: number
  dayProgressPercent: number
  duskProgressPercent: number

  sunAboveHorizon: boolean
  sunPercentOverHorizon: number
  sunPosition: {
    x: number
    y: number
  }
}

export enum EHorizonCardErrors {
  SunIntegrationNotFound = 'SunIntegrationNotFound'
}

export type THorizonCardTimes = {
  dawn?: Date
  dusk?: Date
  noon: Date
  sunrise?: Date
  sunset?: Date
}

export type THorizonCardData = {
  azimuth: number
  elevation: number

  sunInfo: TSunInfo

  times: THorizonCardTimes

  error?: EHorizonCardErrors
}

export enum EHorizonCardI18NKeys {
  Azimuth = 'azimuth',
  Dawn = 'dawn',
  Dusk = 'dusk',
  Elevation = 'elevation',
  Noon = 'noon',
  Sunrise = 'sunrise',
  Sunset = 'sunset'
}

export type THorizonCardI18N = Record<string, unknown>

export type THorizonCardI18NErrorKeys = {
  [key in EHorizonCardErrors]: string
}

export type THorizonCardI18NKeys = { [key in EHorizonCardI18NKeys ]: string } | { errors: THorizonCardI18NErrorKeys }
