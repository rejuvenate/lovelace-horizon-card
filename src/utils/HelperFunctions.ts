import { html, nothing, TemplateResult } from 'lit'

import { Constants } from '../constants'
import { EHorizonCardI18NKeys, TMoonPhase } from '../types'
import { I18N } from './I18N'

type FieldValue = Date | number | string | undefined

export class HelperFunctions {
  public static renderFieldElements (i18n: I18N, translationKey: string, values: FieldValue[],
                                     extraClasses: string[] = []): TemplateResult {
    const mappedValues = values
      .map((value, index) => this.valueToHtml(i18n, translationKey, value, extraClasses[index]))

    return this.renderFieldElement(i18n, translationKey, mappedValues)
  }

  public static renderFieldElement (i18n: I18N, translationKey: string, value: FieldValue | TemplateResult[]): TemplateResult {
    return html`
      <div class="horizon-card-text-container">
        <div class="horizon-card-field-name">${ i18n.tr(translationKey) }</div>
        ${value instanceof Array ? value : this.valueToHtml(i18n, translationKey, value) }
      </div>
    `
  }

  public static renderMoonElement (i18n: I18N, phase: TMoonPhase, phaseRotation: number) {
    if (phase === undefined) {
      return nothing
    }

    let moon_phase_localized: unknown = i18n.localize(`component.moon.entity.sensor.phase.state.${phase.state}`)
    if (!moon_phase_localized) {
      moon_phase_localized = html`<span title="Install Moon integration to get localized Moon phase name">${phase.state} (!)</span>`
    }

    return html`
      <div class="horizon-card-text-container">
        <div class="horizon-card-field-moon-phase" style="transform: rotate(${phaseRotation}deg)">
          <ha-icon icon="mdi:${phase.icon}"></ha-icon>
        </div>
        <div class="horizon-card-field-value horizon-card-field-value-moon-phase">${moon_phase_localized}</div>
      </div>
    `
  }

  private static valueToHtml (i18n: I18N, translationKey: string, value: FieldValue, klass='') {
    const mappedValue = this.fieldValueToString(i18n, translationKey, value)
    return html`<div class="horizon-card-field-value ${klass}">${mappedValue}</div>`
  }

  private static fieldValueToString (i18n: I18N, translationKey: string, value: FieldValue) {
    let pre = ''
    let post = ''
    if (value === undefined) {
      value = '-'
    } else if (value instanceof Date) {
      value = i18n.formatDateAsTime(value)
      const parts = value.match(/(.*?)(\d{1,2}[:.]\d{2})(.*)/)
      if (parts != null) {
        pre = parts[1]
        value = parts[2]
        post = parts[3]
      }
    } else if (typeof value === 'number') {
      value = i18n.formatDecimal(value)
      if (translationKey === EHorizonCardI18NKeys.Azimuth || translationKey === EHorizonCardI18NKeys.Elevation) {
        value += 'Â°'
      }
    }
    const preHtml = pre ? html`<span class="horizon-card-field-value-secondary">${pre}</span>` : nothing
    const postHtml = post ? html`<span class="horizon-card-field-value-secondary">${post}</span>` : nothing

    return html`${preHtml}${value}${postHtml}`
  }

  public static isValidLanguage (language: string): boolean {
    return Object.keys(Constants.LOCALIZATION_LANGUAGES).includes(language)
  }

  public static clamp (min: number, max: number, value: number): number {
    if (min === max) {
      return min
    }

    if (min > max) {
      throw new RangeError('Min value can not be bigger than the max value')
    }

    return Math.min(Math.max(value, min), max)
  }

  public static rangeScale (minRange: number, maxRange: number, range: number, value: number) {
    const clamped = HelperFunctions.clamp(minRange, maxRange, range) - minRange
    const rangeSize = maxRange - minRange
    return (1 - clamped / rangeSize) * value
  }

  public static noonAtTimeZone (date: Date, timeZone: string): Date {
    let tzDate
    try {
      tzDate = this.getTimeInTimeZone(date, '12:00:00', timeZone)
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e)
      tzDate = new Date(date)
      tzDate.setHours(12)
      tzDate.setMinutes(0)
      tzDate.setSeconds(0)
      tzDate.setMilliseconds(0)
    }
    return tzDate
  }

  public static midnightAtTimeZone (date: Date, timeZone: string): Date {
    let tzDate
    try {
      tzDate = this.getTimeInTimeZone(date, '00:00:00', timeZone)
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e)
      tzDate = new Date(date)
      tzDate.setHours(0)
      tzDate.setMinutes(0)
      tzDate.setSeconds(0)
      tzDate.setMilliseconds(0)
    }
    return tzDate
  }

  private static getTimeInTimeZone (date: Date, time: string, timeZone: string) {
    const formatter = new Intl.DateTimeFormat('fr-CA',
      { timeZone: timeZone, timeZoneName: 'longOffset' })
    // 'fr-CA' locale formats like '2023-04-11 UTC+03:00' or '2023-04-11 UTC-10:00' or '2023-04-11 UTC'
    const formatted = formatter.format(date)
    const parts = formatted
      .replace('\u2212', '-') // minuses might be U+2212 instead of plain old ASCII hyphen-minus
      .split(' ')
    let tz = parts[1].replace('UTC', '')
    if (tz === '') {
      tz = 'Z'
    }
    const dateToParse = `${parts[0]}T${time}${tz}`
    const result = new Date(dateToParse)
    if (isNaN(result.getTime())) {
      // Something went fishy with using the above method - generally should not happen
      throw new Error(`Could not convert time to time zone: ${formatted} -> ${dateToParse}`)
    }
    return result
  }
}
