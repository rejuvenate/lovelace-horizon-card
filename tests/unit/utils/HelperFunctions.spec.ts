import { NumberFormat, TimeFormat } from 'custom-card-helpers'

import {Constants} from '../../../src/constants'
import { HelperFunctions } from '../../../src/utils/HelperFunctions'
import { I18N } from '../../../src/utils/I18N'
import { TemplateResultTestHelper } from '../../helpers/TestHelpers'

describe('HelperFunctions', () => {
  const i18n = new I18N('en', 'UTC', TimeFormat.language, NumberFormat.language,
    (key) => key)

  describe('renderFieldElement', () => {
    it('returns a field element template when the provided value is undefined', async () => {
      const renderFun = () => HelperFunctions.renderFieldElement(i18n, 'sunrise', undefined)

      const html = await TemplateResultTestHelper.renderFunction(renderFun)

      expect(html).toMatchSnapshot()
    })

    it('returns a field element template when the provided value is a date for 24h clock', async () => {
      const i18n = new I18N('en', 'UTC', TimeFormat.twenty_four, NumberFormat.language,
        (key) => key)

      const date = new Date('2021-06-07T21:37:17.812Z')

      const renderFun = () => HelperFunctions.renderFieldElement(i18n, 'sunrise', date)

      const html = await TemplateResultTestHelper.renderFunction(renderFun)

      expect(html).toMatchSnapshot()
    })

    it('returns a field element template when the provided value is a date for 12h clock with post-am/pm', async () => {
      const i18n = new I18N('en', 'UTC', TimeFormat.am_pm, NumberFormat.language,
        (key) => key)

      const date = new Date('2021-06-07T21:37:17.812Z')

      const renderFun = () => HelperFunctions.renderFieldElement(i18n, 'sunrise', date)

      const html = await TemplateResultTestHelper.renderFunction(renderFun)

      expect(html).toMatchSnapshot()
    })

    it('returns a field element template when the provided value is a date for 12 clock with pre-am/pm', async () => {
      const i18n = new I18N('tr', 'UTC', TimeFormat.am_pm, NumberFormat.language,
        (key) => key)

      const date = new Date('2021-06-07T21:37:17.812Z')

      const renderFun = () => HelperFunctions.renderFieldElement(i18n, 'sunrise', date)

      const html = await TemplateResultTestHelper.renderFunction(renderFun)

      expect(html).toMatchSnapshot()
    })

    it('returns a field element template when the provided value is a string', async () => {
      const renderFun = () => HelperFunctions.renderFieldElement(i18n, 'sunrise', 'test')

      const html = await TemplateResultTestHelper.renderFunction(renderFun)

      expect(html).toMatchSnapshot()
    })

    it('returns a field element template when the provided value is a number', async () => {
      const renderFun = () => HelperFunctions.renderFieldElement(i18n, 'sunrise', 9)

      const html = await TemplateResultTestHelper.renderFunction(renderFun)

      expect(html).toMatchSnapshot()
    })

    it('returns a field element template when the provided value is a TMoonPhase', async () => {
      const renderFun = () => HelperFunctions.renderMoonElement(i18n, Constants.MOON_PHASES.fullMoon, 0)

      const html = await TemplateResultTestHelper.renderFunction(renderFun)

      expect(html).toMatchSnapshot()
    })

    it('returns a field element template when Moon integration is missing', async () => {
      // When Moon integration is missing hass.localize() returns an empty string
      const i18n = new I18N('en', 'UTC', TimeFormat.language, NumberFormat.language,
        () => '')
      const renderFun = () => HelperFunctions.renderMoonElement(i18n, Constants.MOON_PHASES.fullMoon, 0)

      const html = await TemplateResultTestHelper.renderFunction(renderFun)

      expect(html).toMatchSnapshot()
    })
  })

  describe('isValidLanguage', () => {
    it('returns true the provided language is supported', () => {
      const result = HelperFunctions.isValidLanguage('es')
      expect(result).toBe(true)
    })

    it('returns false the provided language is not supported', () => {
      const result = HelperFunctions.isValidLanguage('notSupported')
      expect(result).toBe(false)
    })
  })

  describe('clamp', () => {
    it('returns the correct value when min and max have the same value', () => {
      const result = HelperFunctions.clamp(50, 50, 20)
      expect(result).toBe(50)
    })

    it('throws an error when min is greater than max', () => {
      let resultError
      try {
        HelperFunctions.clamp(50, 0, 20)
      } catch (error) {
        resultError = error
      }

      expect(resultError).toBeDefined()
      expect(resultError.name).toBe('RangeError')
      expect(resultError.message).toBe('Min value can not be bigger than the max value')
    })

    it('returns the correct value when the value is between min and max', () => {
      const result = HelperFunctions.clamp(0, 50, 20)
      expect(result).toBe(20)
    })

    it('returns the correct value when the value is lower than min', () => {
      const result = HelperFunctions.clamp(40, 50, 20)
      expect(result).toBe(40)
    })

    it('returns the correct value when the value is higher than max', () => {
      const result = HelperFunctions.clamp(40, 50, 60)
      expect(result).toBe(50)
    })
  })

  describe('noonAtTimeZone', () => {
    let consoleErrorSpy: jest.SpyInstance
    beforeAll(() => {
      consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation()
    })

    afterEach(() => {
      consoleErrorSpy.mockClear()
    })

    afterAll(() => {
      consoleErrorSpy.mockRestore()
    })

    it('returns noon in the requested time zone', () => {
      const date = new Date('2023-04-13T23:10:45.532Z')
      const result = HelperFunctions.noonAtTimeZone(date, 'Europe/Sofia')

      expect(result).toEqual(new Date('2023-04-14T12:00:00+03:00'))
      expect(consoleErrorSpy).not.toHaveBeenCalled()
    })

    it('returns noon in local timezone if an error occurs', () => {
      const date = new Date('2023-04-13T23:10:45.532Z')
      const result = HelperFunctions.noonAtTimeZone(date, 'Europe/XXX')

      // We can only verify the time - the date will be +/- 1 according to the actual time zone
      expect(result.getHours()).toEqual(12)
      expect(result.getMinutes()).toEqual(0)
      expect(result.getSeconds()).toEqual(0)
      expect(result.getMilliseconds()).toEqual(0)
      expect(consoleErrorSpy).toHaveBeenCalledTimes(1)
    })
  })

  describe('midnightAtTimeZone', () => {
    let consoleErrorSpy: jest.SpyInstance
    beforeAll(() => {
      consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation()
    })

    afterEach(() => {
      consoleErrorSpy.mockClear()
    })

    afterAll(() => {
      consoleErrorSpy.mockRestore()
    })

    it('returns midnight in the requested time zone', () => {
      const date = new Date('2023-04-13T23:10:45.532Z')
      const result = HelperFunctions.midnightAtTimeZone(date, 'Europe/Sofia')

      expect(result).toEqual(new Date('2023-04-14T00:00:00+03:00'))
      expect(consoleErrorSpy).not.toHaveBeenCalled()
    })

    it('returns midnight in local timezone if an error occurs', () => {
      const date = new Date('2023-04-13T23:10:45.532Z')
      const result = HelperFunctions.midnightAtTimeZone(date, 'Europe/XXX')

      // We can only verify the time - the date will be +/- 1 according to the actual time zone
      expect(result.getHours()).toEqual(0)
      expect(result.getMinutes()).toEqual(0)
      expect(result.getSeconds()).toEqual(0)
      expect(result.getMilliseconds()).toEqual(0)
      expect(consoleErrorSpy).toHaveBeenCalledTimes(1)
    })
  })
})
