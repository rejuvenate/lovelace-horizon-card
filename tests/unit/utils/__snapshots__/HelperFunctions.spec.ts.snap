// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelperFunctions renderFieldElement returns a field element template when Moon integration is missing 1`] = `
<div class="horizon-card-text-container">
  <div class="horizon-card-field-moon-phase"
       style="transform: rotate(0deg)"
  >
    <ha-icon icon="mdi:moon-full">
    </ha-icon>
  </div>
  <div class="horizon-card-field-value horizon-card-field-value-moon-phase">
    <span title="Install Moon integration to get localized Moon phase name">
      full_moon (!)
    </span>
  </div>
</div>
`;

exports[`HelperFunctions renderFieldElement returns a field element template when the provided value is a TMoonPhase 1`] = `
<div class="horizon-card-text-container">
  <div class="horizon-card-field-moon-phase"
       style="transform: rotate(0deg)"
  >
    <ha-icon icon="mdi:moon-full">
    </ha-icon>
  </div>
  <div class="horizon-card-field-value horizon-card-field-value-moon-phase">
    component.moon.entity.sensor.phase.state.full_moon
  </div>
</div>
`;

exports[`HelperFunctions renderFieldElement returns a field element template when the provided value is a date for 12 clock with pre-am/pm 1`] = `
<div class="horizon-card-text-container">
  <div class="horizon-card-field-name">
    Gündoğumu
  </div>
  <div class="horizon-card-field-value ">
    <span class="horizon-card-field-value-secondary">
      ÖS
    </span>
    9:37
  </div>
</div>
`;

exports[`HelperFunctions renderFieldElement returns a field element template when the provided value is a date for 12h clock with post-am/pm 1`] = `
<div class="horizon-card-text-container">
  <div class="horizon-card-field-name">
    Sunrise
  </div>
  <div class="horizon-card-field-value ">
    9:37
    <span class="horizon-card-field-value-secondary">
      PM
    </span>
  </div>
</div>
`;

exports[`HelperFunctions renderFieldElement returns a field element template when the provided value is a date for 24h clock 1`] = `
<div class="horizon-card-text-container">
  <div class="horizon-card-field-name">
    Sunrise
  </div>
  <div class="horizon-card-field-value ">
    21:37
  </div>
</div>
`;

exports[`HelperFunctions renderFieldElement returns a field element template when the provided value is a number 1`] = `
<div class="horizon-card-text-container">
  <div class="horizon-card-field-name">
    Sunrise
  </div>
  <div class="horizon-card-field-value ">
    9
  </div>
</div>
`;

exports[`HelperFunctions renderFieldElement returns a field element template when the provided value is a string 1`] = `
<div class="horizon-card-text-container">
  <div class="horizon-card-field-name">
    Sunrise
  </div>
  <div class="horizon-card-field-value ">
    test
  </div>
</div>
`;

exports[`HelperFunctions renderFieldElement returns a field element template when the provided value is undefined 1`] = `
<div class="horizon-card-text-container">
  <div class="horizon-card-field-name">
    Sunrise
  </div>
  <div class="horizon-card-field-value ">
    -
  </div>
</div>
`;
